text/001x-battery_improvements.md

```diff
@@ -0,0 +1,293 @@
+      <field type="uint8_t" name="percent_remaining" units="%" invalid="UINT8_MAX">Remaining battery energy. Values: [0-100], UINT32_MAX: field not provided.</field>
+      <field type="uint32_t" name="status_flags" display="bitmask" enum="MAV_BATTERY_STATUS_FLAGS">Fault, health, readiness, and other status indications.</field>
+    </message>
+```
+
+`MAV_BATTERY_STATUS_FLAGS` is a new enum that includes faults, charge state, and battery mode:
+
+```xml
+    <enum name="MAV_BATTERY_STATUS_FLAGS" bitmask="true">
+      <description>Battery status flags for fault, health and state indication.</description>
+      <entry value="1" name="MAV_BATTERY_STATUS_FLAGS_READY_TO_USE">
+        <description>
+          The battery is ready to use (fly).
+          Only set if the battery is safe and ready to fly with (not charging, no critical faults, such as those that would set MAV_BATTERY_STATUS_FLAGS_REQUIRES_SERVICE or MAV_BATTERY_STATUS_FLAGS_BAD_BATTERY, etc.).
+        </description>
+      </entry>
+      <entry value="2" name="MAV_BATTERY_STATUS_FLAGS_CHARGING">
+        <description>
+          Battery is charging.
+          Not ready to use (MAV_BATTERY_STATUS_FLAGS_READY_TO_USE would not be set).
+        </description>
+      </entry>
+      <entry value="4" name="MAV_BATTERY_STATUS_FLAGS_CELL_BALANCING">
+        <description>
+          Battery is cell balancing (during charging).
+          Not ready to use (MAV_BATTERY_STATUS_FLAGS_READY_TO_USE would not be set).
+        </description>
+      </entry>
+      <entry value="8" name="MAV_BATTERY_STATUS_FLAGS_FAULT_CELL_IMBALANCE">
+        <description>
+          Battery cells are not balanced.
+          Not ready to use.
+        </description>
+      </entry>
+      <entry value="16" name="MAV_BATTERY_STATUS_FLAGS_AUTO_DISCHARGING">
+        <description>
+          Battery is auto discharging (towards storage level).
+          Not ready to use (MAV_BATTERY_STATUS_FLAGS_READY_TO_USE would not be set).
+        </description>
+      </entry>
+      <entry value="32" name="MAV_BATTERY_STATUS_FLAGS_REQUIRES_SERVICE">
+        <description>
+          Battery requires service (not safe to fly). 
+          This is set at vendor discretion.
+          It is likely to be set for most faults, and may also be set according to a maintenance schedule (such as age, or number of recharge cycles, etc.).
+        </description>
+      </entry>
+      <entry value="64" name="MAV_BATTERY_STATUS_FLAGS_BAD_BATTERY">
+        <description>
+          Battery is faulty and cannot be repaired (not safe to fly). 
+          This is set at vendor discretion.
+          The battery should be disposed of safely.
+        </description>
+      </entry>
+      <entry value="128" name="MAV_BATTERY_STATUS_FLAGS_PROTECTIONS_ENABLED">
```

Comment:
How would you say if only some of the Protection functions are enabled?

AKA voltages and temperature protections enabled only.

Disabling the current protections while listening for the warnings can be a bit more conservative of a system. As cutting off current could have severe consequences.

Reply:
A question for @dakejahl - my assumption here was that all protections are enabled or disabled. When enabled, if you get a fault you're monitoring for then `MAV_BATTERY_STATUS_FLAGS_FAULT_PROTECTION_SYSTEM` is fired and the FETs turn off the battery. YOu might get the cause of the fault in another fault. If it is not enabled you don't get MAV_BATTERY_STATUS_FLAGS_FAULT_PROTECTION_SYSTEM but you might get other faults.

Upshot MAV_BATTERY_STATUS_FLAGS_FAULT_PROTECTION_SYSTEM  is what tells you that you have lost all power due to fault protection, and MAV_BATTERY_STATUS_FLAGS_PROTECTIONS_ENABLED is more of a status bit.

So the real question is whether you care that particular protections are enabled, or just that any are.

Reply:
>Disabling the current protections while listening for the warnings can be a bit more conservative of a system. As cutting off current could have severe consequences.

>A question for @dakejahl - my assumption here was that all protections are enabled or disabled

In my implementation I am defining "protections disabled" as being the FET control protections that would disable the output FETs. I leave the input FET protections enabled all the time. 

>So the real question is whether you care that particular protections are enabled, or just that any are.

You might, but this might be too much detail and could be implementation dependent. I'm not sure though.

Reply:
@hendjoshsr71 @dakejahl I just read the DS-13 pixhawk standard for smart batteries and from that I infer that the reason we have that MAV_BATTERY_STATUS_FLAGS_PROTECTIONS_ENABLED is to know when we're in "airmode" or not - i.e. its a signal that if you fly and this is set you're potentially stuffed. 

So we don't care which protections are enabled, just that any might be. If you agree, the update would be a note "some protections are enabled"

Reply:
@schwacff

Reply:
My thought is that airmode is a really poor name - that's at least the third overload of the term I've seen just in Pixhawk/PX4. 

This is a case where the spec should be updated.

Reply:
>My thought is that airmode is a really poor name

I agree I think the name could be better. Also since this Mavlink message is supposed to be generic I'd rather leave it as PROTECTIONS_ENABLED since some implementations may or may not share the concept of an "airmode"

Reply:
Yep. I really do think the standard should change here. From my read it isn't even airmode so much as "not in the air mode". Either way, as you say, MAVLink designs should be around generic concepts

Reply:
Airmode is a bit specific to PX4 possibly? Assuming you mean the flight mode but I'm not sure if you do or not :)

I'm fine PROTECTIONS_ENABLED if you document what it means and that it means output protections enabled due to any cause ie current, voltage, temperature, power deviations out of spec.

Reply:
The idea is that output protections are enabled when the vehicle is powered but not in use or in the air. Once the vehicle is armed, or starts moving, you disable the protections to prevent false positives from disabling the output.

Reply:
Hey guys, the way we intend it is that "airmode" (which I agree we should change the mode to "flight mode" or something) is to be the mode that the battery enters before takeoff stays in during flight. This is intended to bypass all possible protections that could cause failure in flight, including things like both charge and discharge cutoff. Charge cutoff should also be deactivated, because if it trips you end up discharging through the body diode of the charge FET, making it super hot and potentially failing. All of these faults however should remain active as warnings so that they notify the user in flight, post flight, and are logged.

Reply:
We're really talking about an armed/disarmed state for the battery. Exactly what that is should be implementation specific. Acknowledging the arguments put forward, I'm a bit hesitant here. Warnings aren't really sufficient: I think we need to be clearer that these errors should trigger failsafe actions, but that whilst armed the failsafe action will be controlled by the autopilot not the battery.

Reply:
I agree - the choice of what to do about these should be up to the application / aircraft

Reply:
@auturgy The semantics of "armed" are different enough I think that using this terminology is risky. Unlike with autopilot armed (in which power is not supplied), the battery will normally still supply power unless there is a fault. Further the protection system might still detect a fault but not trigger the failsafe.

But I do very much like the idea that this is a mode where all faults should be ignored by the autopilot while in the air _if we can make that assumption_ - @schwacff ?

In that case we might need a better name than armed or protection. Perhaps MAV_BATTERY_STATUS_FLAGS_BATTERY_CONTROLS_FAILSAFE ?

Reply:
> Airmode is a bit specific to PX4 possibly? Assuming you mean the flight mode but I'm not sure if you do or not :)

@hendjoshsr71 Yes, airmode in this case means "safe to fly because the protection system is not enabled". Not the PX4 [airmode](http://docs.px4.io/master/en/config_mc/pid_tuning_guide_multicopter.html#airmode-mixer-saturation). I'm OK with protections as a generic term, but we might be heading to even better semantics with ^^^.

Reply:
I'm not sold on the "failsafe" term. That is used more to indicate the opposite of what we need, which is in case of failure, keep running. I like "Flight Mode"!

Reply:
@schwacff Well, that's the inversion. Either way I hate "flight mode" with real passion because its a well defined term with a _different_ meaning. Ditto airmode

Maybe someone else can come up with a better option.

Reply:
> The idea is that output protections are enabled when the vehicle is powered but not in use or in the air. Once the vehicle is armed, or starts moving, you disable the protections to prevent false positives from disabling the output.

Thanks Alex! This explains what I hoped you meant by it. Add in some of the additional details from the Cory and  anew name and it should make sense to a knowledgeable reader.

Reply:
@hendjoshsr71 OK, I have added a note for the flag capturing the above comment: https://github.com/mavlink/rfcs/pull/19/files#r976062357

I haven't changed the name because I don't think we have a better one yet.

Reply:
Yeah you are right on the name.   I couldn't determine a better one so I Iet it be the last time i looked through this PR.

 I like the additional wording for disabling protections when armed.